Fix double dispatch to Python for detach

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163671

This fixes #71725.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

swolchok committed:
Fix double dispatch to Python for detach?

I am not very sure what I'm doing here, but this appears to fix #71725. I would very much appreciate feedback about whether this is the right thing to do here.

I manually ran the repro.py attached to #71725 to verify the fix, as well as noting that DTensor.detach got faster. I would also appreciate feedback about where a conversion of that repro into a regression test should live, because I am not particularly familiar with autograd or view ops broadly.

[ghstack-poisoned]

swolchok committed:
fix one test, resnapshot others on "Fix double dispatch to Python for detach?"


I am not very sure what I'm doing here, but this appears to fix #71725. I would very much appreciate feedback about whether this is the right thing to do here.

I manually ran the repro.py attached to #71725 to verify the fix, as well as noting that DTensor.detach got faster. ~~I would also appreciate feedback about where a conversion of that repro into a regression test should live, because I am not particularly familiar with autograd or view ops broadly.~~ I will update TestPythonDispatch::test_detach_appears_twice_when_called_once to be a regression test.

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -450,6 +450,17 @@ static const Tensor& resize_as_(
 static Tensor detach(c10::DispatchKeySet ks, const Tensor& self) {
   auto out = ([&]() {
     at::AutoDispatchBelowADInplaceOrView guard;
+    // HACK: avoid doubling cost if we're going to dispatch to Python,
```
So, one problem with doing this is detach() will never get called at all if base.is_inference(), and this is undesirable:

```
  // Note [View of inference tensor]
  // For inference tensor this code can only be hit outside InferenceMode
  // since ADInplaceOrView is in the default_included_set.
  // If Inplace and View were separate dispatch keys we can just put Inplace
  // in the default_included_set, so that view ops on inference tensor doesn't
  // have to go through as_view even outside InferenceMode.
  if (base.is_inference())
    return tensor;
```

ezyang replied:
The key operation we do in as_view is this:

```
// See NOTE [ Autograd View Variables ] for details.
// Non-differentiable view. Just share version counter.
inline Variable make_variable_non_differentiable_view(
    const Variable& base,
    const at::Tensor& data,
    bool allow_tensor_metadata_change = true) {
  if (data.defined()) {
    // Currently all of non-differentiable view ops(detach/_indices/_values)
    // share the same TensorImpl as their base Tensor. Thus a new TensorImpl
    // allocation here is required.
    auto data_impl_copy = data.getIntrusivePtr()->shallow_copy_and_detach(
        /*version_counter=*/impl::version_counter(base),
        /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
    data_impl_copy->set_autograd_meta(nullptr);
    return Variable(data_impl_copy);
  }
  return Variable();
}
```

The payload for detach() in normal code is:

```
Tensor detach(const Tensor& self) {
  // NB: detach() is not the same thing as alias()! The main difference is that
  // detach does not allow metadata change while alias does.
  return Tensor(self.getIntrusivePtr()->shallow_copy_and_detach(
      // NB: The ADInplaceOrView logic will overwrite these with the
      // appropriate values if it runs; otherwise these are the values.
      /*version_counter=*/0,
      /*allow_tensor_metadata_change=*/false));
}
```

But shallow_copy_and_detach on subclass does this:

```
template <typename VariableVersion>
c10::intrusive_ptr<TensorImpl> TensorImpl::shallow_copy_and_detach_core(
    VariableVersion&& version_counter,
    bool allow_tensor_metadata_change) const {
  c10::intrusive_ptr<TensorImpl> r;
  const auto mode_stack_len = c10::impl::TorchDispatchModeTLS::stack_len();
  // TODO: do we have to exclude after Python dispatch key set?
  if (mode_stack_len > 0 &&
      !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
    const auto& cur_torch_dispatch_mode_state =
        c10::impl::TorchDispatchModeTLS::get_stack_at(mode_stack_len - 1);
    r = cur_torch_dispatch_mode_state->pyinterpreter()->detach(this);
  } else if (
      key_set_.has(DispatchKey::Python) &&
      !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
    r = (pyobj_slot_.load_pyobj_interpreter())->detach(this);
  }
  if (r) {
    if (!r->is_inference()) {
      r->set_version_counter(std::forward<VariableVersion>(version_counter));
    }
    r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
    return r;
  }
  // otherwise just copy the TensorImpl and not the PyObject.  Since
  // the interpreter is dead no one can call us out on it
  auto impl = c10::make_intrusive<TensorImpl>(
      // No need to populate Storage; copy_tensor_metadata will do it for us.
      key_set_,
      data_type_,
      device_opt_);
  copy_tensor_metadata(
      /*src_impl=*/this,
      /*dest_impl=*/impl.get(),
      /*version_counter=*/std::forward<VariableVersion>(version_counter),
      /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
  return impl;
}
```

I think my spicy proposal is that we should just NEVER redispatch to the underlying detach. It seems pointless.

@albanD wdyt?

albanD replied:
> so the only thing it's doing is set the autograd meta to nullptr, right?

Yes.

> would that be a desirable cleanup from your perspective?

It is all quite convoluted one way or the other....
But keeping this one next to make_variable_differentiable_view (that has the same assumptions) feels ok to me (and less refactor work for you :D )

ezyang reviewed (approved):
OK with the dog science version too

swolchok committed:
switch to Alban's proposal -- remove shallow_copy_and_detach. resnapshot a couple more tests on "Fix double dispatch to Python for detach?"


I am not very sure what I'm doing here, but this appears to fix #71725. I would very much appreciate feedback about whether this is the right thing to do here.

I manually ran the repro.py attached to #71725 to verify the fix, as well as noting that DTensor.detach got faster. ~~I would also appreciate feedback about where a conversion of that repro into a regression test should live, because I am not particularly familiar with autograd or view ops broadly.~~ I will update TestPythonDispatch::test_detach_appears_twice_when_called_once to be a regression test.

[ghstack-poisoned]

swolchok committed:
forgot to amend, sigh on "Fix double dispatch to Python for detach?"


I am not very sure what I'm doing here, but this appears to fix #71725. I would very much appreciate feedback about whether this is the right thing to do here.

I manually ran the repro.py attached to #71725 to verify the fix, as well as noting that DTensor.detach got faster. ~~I would also appreciate feedback about where a conversion of that repro into a regression test should live, because I am not particularly familiar with autograd or view ops broadly.~~ I will update TestPythonDispatch::test_detach_appears_twice_when_called_once to be a regression test.

[ghstack-poisoned]

albanD reviewed (commented):
.

swolchok committed:
Update on "Fix double dispatch to Python for detach"


I am not very sure what I'm doing here, but this appears to fix #71725. I would very much appreciate feedback about whether this is the right thing to do here.

I manually ran the repro.py attached to #71725 to verify the fix, as well as noting that DTensor.detach got faster. ~~I would also appreciate feedback about where a conversion of that repro into a regression test should live, because I am not particularly familiar with autograd or view ops broadly.~~ I will update TestPythonDispatch::test_detach_appears_twice_when_called_once to be a regression test.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok commented:
@albanD looks like your proposal didn't work? https://github.com/pytorch/pytorch/actions/runs/18144312914/job/51645283222?pr=163671#step:25:1353

swolchok committed:
use impl::version_counter; expect this to still be broken on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok committed:
fix SparseTensor invariant violation thanks to Alban on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok commented:
@swolchok has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D83857880).

swolchok commented:
new sparse test failure (python test/test_torch.py TestTorchDeviceTypeCPU.test_grad_scaling_unscale_sparse_cpu_float32 assertion failure on new_nnz <= nnz()) is definitely not associated with an ASAN problem, double-checked

swolchok committed:
execute bailout plan on "Fix double dispatch to Python for detach/_indices/_values"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

swolchok committed:
move if statement below comment and fix mistake in it on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok committed:
resnapshot 3 more tests on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
fix even more test_memory_profiler failures on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "breaks export tests" -c nosignal

export/test_export_training_ir_to_run_decomp.py::TrainingIRToRunDecompExportNonStrictTestExport::test_tag_ac_export_training_ir_to_decomp_nonstrict [GH job link](https://github.com/pytorch/pytorch/actions/runs/18329076162/job/52202433523) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/c32118dc3e50505fd285e6e448a90883fce11535)

export/test_export.py::TestExport::test_tag_ac_export [GH job link](https://github.com/pytorch/pytorch/actions/runs/18329076162/job/52202692797) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/c32118dc3e50505fd285e6e448a90883fce11535)


Looks like the test expectations need to be updated.

swolchok committed:
rebase, resnapshot test_export.py on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
resnapshot two more failing tests on "Fix double dispatch to Python for detach"


This fixes #71725.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
rebase on "Fix double dispatch to Python for detach"


This fixes #71725.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
resnapshot again, not sure why it didn't take last time on "Fix double dispatch to Python for detach"


This fixes #71725.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
yet another rebase and resnapshot on "Fix double dispatch to Python for detach"


This fixes #71725.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok committed:
rebase again; had the bad luck of rebasing atop of a diff that broke a bunch of tests on "Fix double dispatch to Python for detach"


This fixes #71725.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

Differential Revision: [D83857880](https://our.internmc.facebook.com/intern/diff/D83857880)

[ghstack-poisoned]

swolchok commented:
looking at HUD, https://github.com/pytorch/pytorch/actions/runs/18484981746/job/52668577498?pr=163671 is a KP from yesterday that was fixed later. Merging with -i.
